### DISCUSSION TITLE
Why is it taking 10s+ for a hot reload for a small change? 😞

### DISCUSSION THREAD

**Discussion by abhagsain** [vercel/next.js#12831] [Help] [answered]
I'm not that much experienced working with Next.js and I'm still learning it. 
But this slow hot reload is making it even more harder for me to build something. 
And by a small change I mean moving one line up or down. 

### Here's a demo

![timer](https://user-images.githubusercontent.com/36589645/81881687-70f25800-95ae-11ea-8935-0566b872d671.gif)

Somebody please help me resolve this issue. This is my [repository](https://github.com/abhagsain/jsmtc/tree/master)

Thank you!

### Edit (14 May 2020)

Since my codebase is small I'm moving it to Create React App. I wanted to build a project and learn Next.js but the slow load time is holding me back from getting anything done.
I know it's not good to switch technologies when encountered a bug but I have already wasted 3 days trying to fix it and it's my personal project so it's okay for me 🤷‍♂️

Thank you everyone for your time.

**Comment by helloitsm3** [comment]
I'm having the same issue as well. After updating to the latest version, I'm having to wait quite a while whenever I'm making a small change. The change could be as small as changing a text and it's taking so long to reload

**Comment by VivekNeel** [comment]
I'm having the same problem as well. They said they have FastRefresh now but that does not seems to work. I'm on next 9.4.0

**Comment by brunocrosier** [comment]
@abhagsain I cloned your repo locally and it reloaded normally on my machine. Here is a video: https://streamable.com/07gria

What happens when you do `npx create-next-app` and start from scratch? Does the reloading spead improve?

**Reply by riccardogiorato** [reply]
I have had your same speed on this project.I can't reproduce this slowliness without more details.

**Reply by Timer** [reply]
Could you please provide your machine specs so we can compare to OPs?

**Reply by prevostc** [reply]
Maybe the nodejs version used can be very outdated and lead to build speed issues (`node --version`)

**Reply by brunocrosier** [reply]
My specs: 

![image](https://user-images.githubusercontent.com/18399089/81815256-adc43d80-9521-11ea-85be-3f008eec26b1.png)

**Reply by helloitsm3** [reply]
> 
> 
> Maybe the nodejs version used can be very outdated and lead to build speed issues (`node --version`)

What's your node version? I'm running node v13.12.0

**Reply by brunocrosier** [reply]
> > Maybe the nodejs version used can be very outdated and lead to build speed issues (`node --version`)
> 
> What's your node version? I'm running node v13.12.0

I'm on 12.14.1

**Reply by VivekNeel** [reply]
Video link is broken FYI

**Reply by brunocrosier** [reply]
no video definitely works, tested in incognito @VivekNeel

**Reply by helloitsm3** [reply]
> 
> 
> Video link is broken FYI

Just tested on my side too. Video is working fine

**Comment by Timer** [comment]
Could you please provide your machine specs?

**Reply by helloitsm3** [reply]
![image](https://user-images.githubusercontent.com/13825491/81816956-9a839300-955e-11ea-960d-9e64f208218e.png)

Edit: I highly doubt that this issue has to do with my machine specs. I was able to run very smoothly with the previous version of NextJS but the moment I upgraded to the latest version, I'm having very slow hot reloading time.

**Reply by brunocrosier** [reply]
@helloitsm3 does this happen on a clean `npx create-next-app` ?

**Reply by helloitsm3** [reply]
Yes, I've tried it and it's taking really long to reload.

**Reply by brunocrosier** [reply]
Seems like we're on different node versions. Maybe you could try and downgrade to 12.16.3 (most up-to-date LTS version) and see if that improves it?

**Reply by helloitsm3** [reply]
> 
> 
> Seems like we're on different node versions. Maybe you could try and downgrade to 12.16.3 (most up-to-date LTS version) and see if that improves it?

But it doesn't make any sense if my NextJS is able to load fast in the previous version but not the latest. I guess I'll downgrade my NextJS instead. Thanks for your help.

**Reply by abhagsain** [reply]
> Seems like we're on different node versions. Maybe you could try and downgrade to 12.16.3 (most up-to-date LTS version) and see if that improves it?

@brunocascio Hi Bruno. I use nvm and I was on ```node 10.19.0``` and no luck even when I upgraded to ```node 12.16.0``` 😪

### Here's a demo with a stopwatch

![timer](https://user-images.githubusercontent.com/36589645/81843641-85613100-956b-11ea-938a-f165d1d979a7.gif)

### My Current Node Version 

![version](https://user-images.githubusercontent.com/36589645/81844131-41226080-956c-11ea-8d05-07700227d925.png)


This is my system information

![system](https://user-images.githubusercontent.com/36589645/81841734-80e74900-9568-11ea-933c-5f93fd88389b.png)

**The only thing left to try is creating a new project on** ```node 12.16.3```  🤷‍♂️

**Reply by abhagsain** [reply]
@timer @brunocrosier 

### Here's a demo if I create a new project from ```npx create-next-app```

![newproject](https://user-images.githubusercontent.com/36589645/81880936-6c2ca480-95ac-11ea-84ff-d6a7b6dca0b1.gif)

I don't know what's wrong 😭

**Reply by helloitsm3** [reply]
> 
> 
> @Timer @brunocrosier
> ### Here's a demo if I create a new project from `npx create-next-app`
> 
> ![newproject](https://user-images.githubusercontent.com/36589645/81880936-6c2ca480-95ac-11ea-84ff-d6a7b6dca0b1.gif)
> 
> I don't know what's wrong 😭

I'm having the same issue as you and I tried many ways to fix it but to no avail. I don't think it has anything to do with the hardware specs. I've decided to downgrade my NextJS to version 9.3.6 and it is working great now. Able to compile much faster even on small changes. I suggest you to do the same until there's a fix for this issue.

**Reply by abhagsain** [reply]
@helloitsm3 At first it loaded in 4s but now it's again taking ~13 seconds to load even on 9.3.6. It's my fault I should have gone with CRA 🙄

**Reply by helloitsm3** [reply]
> 
> 
> @helloitsm3 At first it loaded in 4s but now it's again taking ~13 seconds to load even on 9.3.6. It's my fault I should have gone with CRA 🙄

Mine takes less than a second to load on 9.3.6. It's really weird that you're experiencing longer reloading time

**Comment by Timer** [comment]
Could everyone please try `next@^9.4.1-canary.5` (or newer) and see if that resolves their speed issue?

**Reply by mondoadams** [reply]
My Fast Refresh is still running very slow, even after upgrading to `next@9.4.1-canary.5`. I'm on Windows using WSL 2. Is there anything else to try?

**Comment by sambhav-gore** [comment]
I am experiencing some slowness as well. I upgraded from 9.2.x to 9.4.0 
but it is faster for Production build.

@Timer can I ship on Canary branch ?

**Reply by sambhav-gore** [reply]
Not Branch, I meant Version

**Reply by Timer** [reply]
Yes, please upgrade to `next@9.4.1-canary.5`.

**Reply by sambhav-gore** [reply]
I will try it now. Just FYI after I upgraded to 9.4.0 I have some static images in the /public dir and on production build serving those images has become extremely slow. 
I plan to move them to a CDN anyway, but I wonder why it is so much slower now for me. 
note that I cannot deploy on Now platform due to some constraints with the domain/hosting right now, and my app is deployed on a K8s elsewhere. before the 9.4.0 upgrade it was fine.

**Reply by michaelwitk** [reply]
@Timer 🤗 my refresh times are <=3 seconds again, thank you all so much!

(using `next@9.4.1-canary.5`)
(Looked through the commit, and yes I am a windows user 👍 )

**Reply by sambhav-gore** [reply]
My refresh time is lightning fast as well after going to 9.4.1-canary.5 
but I am facing a different challenge now as described above. posted here - https://github.com/zeit/next.js/discussions/12932

**Reply by helloitsm3** [reply]
> 
> 
> Yes, please upgrade to `next@9.4.1-canary.5`.

Thanks for sharing this but it is somehow leading to another problem with static images. Any fix to it?

**Comment by raitucarp** [comment]
My case was, didn't hot reloading/refresh.

**Reply by raitucarp** [reply]
I use WSL Ubuntu with node v12.16.2 and yarn

**Reply by raitucarp** [reply]
I have confirmed that switch to run on windows solves my issue. WSL didn't work with 9.4.0. 
However, I didn't measure the time yet.

**Reply by abhijeet-ixr** [reply]
Facing the same with WSL 2  and Ubuntu 20.04 LTS on Next v9.4.4

**Reply by krishnatejakanchi** [reply]
Same situation here.

**Reply by helloitsm3** [reply]
I'm on the latest NextJS @ v9.4.4 and it seems to be quite fast now. You should try to create a new project and try to see if there's same issue

**Reply by abhijeet-ixr** [reply]
@helloitsm3 : The issue in this comment thread is around NextJS on WSL. I have tested out v9.4.4 on WSL and it totally doesn't reload on changes. It works like a charm when using over just windows and not involving WSL. I had to move from WSL based dev env to just windows.

So did you test v9.4.4 over Windows+WSL or just Windows?

**Reply by helloitsm3** [reply]
> 
> 
> @helloitsm3 : The issue in this comment thread is around NextJS on WSL. I have tested out v9.4.4 on WSL and it totally doesn't reload on changes. It works like a charm when using over just windows and not involving WSL. I had to move from WSL based dev env to just windows.
> 
> So did you test v9.4.4 over Windows+WSL or just Windows?

I tested on just Windows. I've not try it on Windows + WSL

**Reply by sosalejandro** [reply]
I'm also a WSL2 Ubuntu 20.04 user and there are quite a few reports on the web about how slow it performs on writing files. It might be the issue, reading/writing isn't optimized, it's performance is very poor.

**Reply by DannyFeliz** [reply]
I'm facing the same issue, here is my discussion https://github.com/vercel/next.js/discussions/22214

**Comment by DittoLabsNg** [comment, marked as answer]
The slow reload in my project was due to my tailwindcss file structure. So imported the tailwind styles and the my normal stylesheet separately. This solved it

**Reply by GeniusHawlah** [reply]
This comment helped me but I am not sure I did it the way you did. 

Initially, I used the @tailwind to import my  tailwind in my global css folder which makes slow. Now I removed it and it works fast but cant apply tailwind classes inline.

**Reply by AntCas** [reply]
@GeniusHawlah have you found a solution that lets you still use Tailwind?

**Reply by AntCas** [reply]
This fixed it for me: https://stackoverflow.com/a/63251133

TL;DR, upgrade to latest Tailwind and enable their new JIT compiler.

**Comment by alirezanasseh** [comment]
I have the same problem, did you find the solution?

**Reply by helloitsm3** [reply]
How long did yours take to reload?

**Reply by alirezanasseh** [reply]
About 30s, I use Next.js 10 on Windows 10. I checked my project folder is excluded in Windows Defender.

**Reply by rockgabi** [reply]
This still happens in 2023. Sometimes randomly I need to wait more than 40 secs. Sometimes it won't even finish reloading.

**Comment by paulussimanjuntak** [comment]
Is there any solution for this problem?

**Comment by en3sis** [comment]
I had the same issue with`"next": "10.0.6",` and Tailwindcss and by the time I was writing this comment I noticed something:

At the end [of the guide](https://tailwindcss.com/docs/guides/nextjs) it recommends (and if you don't read carefully you might be doing both options) importing as ` import "tailwindcss/tailwind.css"` in your `_app.js` or with the `@tailwind....` in your `global.css`. 
I was doing both....🤦🏼‍♂️  by removing the `@tailwind...` import from my `global.css` file, all performance issues are gone now.

**Reply by niteshkumar2000** [reply]
Thanks! I was doing the same mistake xD, you saved me so much time

**Reply by gorfadvijay** [reply]
me too thanks

**Reply by Queen-esther01** [reply]
I am ashamed to say i also made the same mistake. @en3sis thank you XD

**Reply by nwazuo** [reply]
I think this may have been the case for me as well as I things are much faster after removing `tailwindcss/tailwind.css` from `_app.js`

**Comment by runoncedev** [comment]
I noticed that this only happens when I modify the CSS file (https://github.com/oncet/coel/blob/main/styles/globals.css). If I modify a page or a component, times are a lote faster.

**Reply by milan090** [reply]
If you separate this into 2 CSS files: one for the first 3 tailwind lines, and the other for rest of the css, it'll be much faster as it doesn't have to reload the large tailwind css everytime

**Reply by runoncedev** [reply]
Thankfully with the "just-in-time" this is no longer a problem. Everything works fast and smooth.

**Comment by AntCas** [comment]
Here's the fix I found: https://stackoverflow.com/a/63251133

- upgrade to latest Tailwind ^2.1
- enable new JIT mode to improve hot-reload speeds.

**Comment by kopyl** [comment]
Just got the same issue on a top MacBook.
Rebooting my laptop helped. Restarting yarn did not.
Guys, please fix this @Timer

**Reply by wolfitoXtreme** [reply]
I'm experiencing the same, with NextJS 14, in more that one project with different dependencies each, seems is not only related to tailwind as I am not using that thing, rather it seems related to the non CSS modules stylesheets, (at least for me, I'm using Sass), or how they are compiled, DOM changes and CSS modules are quick to appear but whenever I update a single style that is global it takes quite a while to update. It is frustrating, perhap I should consider use another framework and Vite instead.

**Reply by wolfitoXtreme** [reply]
I'm using React MUI (Base which is extra small) in one of these projects and it is even worse, not even DOM changes are fast there.

**Reply by wolfitoXtreme** [reply]
OK, seems the antivirus is to blame this time
https://stackoverflow.com/questions/78130155/why-is-hot-reloading-extremely-slow-in-my-next-js-14-project-with-tailwindcss-s

**Comment by BodhiHu** [comment]
restarted the dev server and hmr is fast again (~5s):
```
[Fast Refresh] done in 5628ms
hot-reloader-client.js:187 [Fast Refresh] rebuilding
hot-reloader-client.js:44 [Fast Refresh] done in 3780ms
```

noticed another thing, next-server is consuming lots of memory.

**Comment by peterburgs** [comment]
I'm using Next.js version 15.5.6 and it takes me about 10-30 seconds to update a small change (even when remove a single charater).
I tried to remove .next, node_modules, restart dev server,clear cache.... but nothing get better.
My final decision is to use --turbopack:
`"dev": "next dev --turbopack",`
with this new option, it takes just 1-2 second or even less.
